fix(lua transform): support sub-second timestamp resolution

jschaf opened the pull request:
Add an optional `nanosecond` field to the `FromLua` code to support
sub-second resolution in timestamps.

The default Lua table format for timestamps has an integer seconds
field. There's no sub-second field like milliseconds in the table
timestamp format. This means it's not possible to send timestamps
with sub-second resolution from Lua to Vector using the timestamp
table format alone.

Closes #2332.

jschaf commented on src/transforms/lua/v2/interop/util.rs:
```diff
@@ -35,7 +33,8 @@ pub fn table_to_timestamp<'a>(t: LuaTable<'a>) -> LuaResult<DateTime<Utc>> {
     let min = t.get("min")?;
     let sec = t.get("sec")?;
-    Ok(Utc.ymd(year, month, day).and_hms(hour, min, sec))
+    let nano = t.get("nanosecond").unwrap_or(0);
```
I'm a rust newbie, not sure what the appropriate idiom is here.

a-rodin replied:
I think it is better to use something like `t.get::<_, Option<u64>>("nanosecond")?.unwrap_or(0)`. The difference is that if a `Result` is not `Ok(...)`, then the error might be caused by reasons other than missing key, such as runtime errors. On the other hand, if there are no runtime errors, but the key is missing, then `get::<_, Option<u64>>?` would result in `None` and then applying `unwrap_or(..)` on an `Option`, as opposed to a result, would not mean ignoring any errors.

Also see https://github.com/timberio/vector/issues/2332#issuecomment-614047567 about the name of the field.

jschaf replied:
Done, thanks for explaining.

jschaf commented on website/metadata.js:
```diff
@@ -64,6 +64,15 @@ module.exports = {
       },
       "description": "Take Vector from zero to production in under 10 minutes.",
       "guides": [
+        {
```
Ran `make generate` because that's what test failure said.

binarylogic replied:
Could you rebase with `master`? This should be resolved.

jschaf replied:
Rebased fixed it.

a-rodin commented:
Should the `nanosec` field also be added when the timestamps are passed from Rust to Lua?

jschaf commented:
> Should the nanosec field also be added when the timestamps are passed from Rust to Lua? Also, the documentation should be updated, see DOCUMENTING.md.

`nanosec` should probably go both ways. I implemented it for both `FromLua` and `ToLua`.  I added documentation with an example of parsing a timestamp format since that's probably what most people need.

jschaf commented:
@binarylogic What's the merge process? I can't merge because I don't have write access.

binarylogic commented:
Hi @jschaf let me get a second set of eyes on the code. @a-rodin is out this week on vacation. We'll try to merge shortly if everything checks out.

LucioFranco commented on website/docs/reference/transforms/lua.md:
```diff
@@ -717,7 +717,7 @@ by the following table:
 | [`Float`][docs.about.data-model.log#floats] | [`number`][urls.lua_number] ||
 | [`Boolean`][docs.about.data-model.log#booleans] | [`boolean`][urls.lua_boolean] ||
-| [`Timestamp`][docs.about.data-model.log#timestamps] | [`table`][urls.lua_table] | There is no dedicated timestamp type in Lua. Because of that timestamps are represented as tables using the convention defined by [`os.date`][urls.lua_os_date] and [`os.time`][urls.lua_os_time]. The table representation of a timestamp  contain fields `year`, `month`, `day`, `hour`, `min`, `sec`, `yday`, `wday`, `i ...
+| [`Timestamp`][docs.about.data-model.log#timestamps] | [`table`][urls.lua_table] | There is no dedicated timestamp type in Lua. Timestamps are represented as tables using the convention defined by [`os.date`][urls.lua_os_date] and [`os.time`][urls.lua_os_time]. The table representation of a timestamp contains the fields `year`, `month`, `day`, `hour`, `min`, `sec`, `yday`, `wday`, and `isdst`. I ...
```
Is this missing a mention to `nanosec`?

binarylogic replied:
I'll update.

binarylogic commented:
Thanks @jschaf!
